pkg/handlers/v1/storage/db.go

```diff
@@ -0,0 +1,309 @@
+	"	aws_resources_id, aws_ips_ip, aws_hostnames_hostname, is_public, is_join, ts, aws_resources.account_id, aws_resources.region, aws_resources.type, aws_resources.meta " +
+	"FROM " +
+	"	" + tableAWSEventsIPSHostnames + " " +
+	"JOIN " + tableAWSResources + " " +
+	"	ON aws_resources_id = aws_resources.id " +
+	"WHERE "
+
+// DB represents a convenient database abstraction layer
+type DB struct {
+	sqldb *sql.DB      // this is a unit test seam
+	LogFn domain.LogFn // also a seam
+
+	once sync.Once
+}
+
+// Init initializes a connection to a Postgres database according to the environment variables POSTGRES_HOST, POSTGRES_PORT, POSTGRES_USER, POSTGRES_PASSWORD, POSTGRES_DATABASE
+func (db *DB) Init(ctx context.Context, postgresConfig *domain.PostgresConfig) error {
+	var initerr error
+	db.once.Do(func() {
+
+		host := postgresConfig.Hostname
+		port := postgresConfig.Port
+		user := postgresConfig.Username
+		password := postgresConfig.Password
+		dbname := postgresConfig.DatabaseName
+
+		logger := db.LogFn(ctx)
+
+		if db.sqldb == nil {
+			sslmode := "disable"
+			if host != "localhost" {
+				sslmode = "enable"
+			}
+			psqlInfo := fmt.Sprintf("host=%s port=%s user=%s "+
+				"password=%s dbname=%s sslmode=%s",
+				host, port, user, password, dbname, sslmode)
+			pgdb, err := sql.Open("postgres", psqlInfo)
+			if err != nil {
+				logger.Error(logs.DBOpenError{Reason: err.Error()})
+				initerr = err
+			}
+
+			err = pgdb.Ping()
+			if err != nil {
+				logger.Error(logs.DBPingError{Reason: err.Error()})
+				initerr = err
+			}
+
+			db.sqldb = pgdb
+		}
+	})
+	return initerr
+}
+
+// StoreCloudAsset an implementation of the Storage interface that records to a database
+func (db *DB) StoreCloudAsset(ctx context.Context, cloudAssetChanges domain.CloudAssetChanges) error {
+	tx, err := db.sqldb.Begin()
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		switch err {
+		case nil:
+			err = tx.Commit()
+		default:
+			err = tx.Rollback()
+		}
+	}()
+
+	if err = db.saveResource(ctx, cloudAssetChanges); err != nil {
+		return err
+	}
+
+	for _, val := range cloudAssetChanges.Changes {
+		err = db.recordNetworkChanges(ctx, cloudAssetChanges.ARN, cloudAssetChanges.ChangeTime, val, tx)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (db *DB) saveResource(ctx context.Context, cloudAssetChanges domain.CloudAssetChanges) error {
+	// You won't get an ID back if nothing was done.  Also, this lib won't return the ID anyway even without the "ON CONFLICT DO NOTHING".
+	// See https://stackoverflow.com/questions/34708509/how-to-use-returning-with-on-conflict-in-postgresql
+	sqlStatement := fmt.Sprintf(`INSERT INTO %s VALUES ($1, $2, $3, $4, $5) ON CONFLICT DO NOTHING RETURNING id`, tableAWSResources)  // nolint
+
+	if _, err := db.sqldb.ExecContext(ctx, sqlStatement, cloudAssetChanges.ARN, cloudAssetChanges.AccountID, cloudAssetChanges.Region, cloudAssetChanges.ResourceType, cloudAssetChanges.Tags); err != nil {
```

Comment:
does this not need to be a part of the transaction as well?

Reply:
Yes it does. Good catch